File: ios/PlaytomicFoundation/logger/Log.swift
//
//  Log.swift
//  PlaytomicFoundation
//
//  Created by Angel Luis Garcia on 10/8/21.
//  Copyright © 2021 Playtomic. All rights reserved.
//
// swiftlint:disable opening_brace

import Foundation
import os

public enum Log {
    public enum Level: Int { case error = 0, warn = 1, info = 2, debug = 3 }

    public static var level: Level = .warn

    public static var errorHandlers: [(_ tag: String, _ message: String) -> Void] = [
        { tag, message in logger(tag: tag).error("\(message)") }
    ]
    public static var warningHandlers: [(_ tag: String, _ message: String) -> Void] = [
        { tag, message in logger(tag: tag).warning("\(message)") }
    ]
    public static var infoHandlers: [(_ tag: String, _ message: String) -> Void] = [
        { tag, message in logger(tag: tag).info("\(message)") }
    ]

    public static func e(_ tag: String, _ messageHandler: @autoclosure () -> String) {
        guard level.rawValue >= Level.error.rawValue else { return }
        let message = messageHandler()
        errorHandlers.forEach { $0(tag, message) }
    }

    public static func w(_ tag: String, _ messageHandler: @autoclosure () -> String) {
        guard level.rawValue >= Level.warn.rawValue else { return }
        let message = messageHandler()
        warningHandlers.forEach { $0(tag, message) }
    }

    public static func i(_ tag: String, _ messageHandler: @autoclosure () -> String) {
        guard level.rawValue >= Level.info.rawValue else { return }
        let message = messageHandler()
        infoHandlers.forEach { $0(tag, message) }
    }

    public static func d(_ tag: String, _ messageHandler: @autoclosure () -> String) {
        guard level.rawValue >= Level.debug.rawValue else { return }
        let message = messageHandler()
        logger(tag: tag).debug("\(message)")
    }

    private static func logger(tag: String) -> Logger {
        Logger(subsystem: Bundle.main.bundleIdentifier ?? "PlaytomicFoundation", category: tag)
    }
}

File: android/app/src/main/java/com/playtomic/test/foundation/logger/Log.kt
package com.playtomic.foundation.logger

object Log {
    enum class Level { Error, Warn, Info, Debug }
    var level: Level = Level.Warn

    val errorHandlers: MutableList<(tag: String, message: String, cause: Throwable?) -> Unit> = mutableListOf(
        { tag, message, cause -> android.util.Log.e(tag, "$message [ $cause ]") }
    )
    val warningHandlers: MutableList<(tag: String, message: String) -> Unit> = mutableListOf(
        { tag, message -> android.util.Log.w(tag, message) }
    )
    val infoHandlers: MutableList<(tag: String, message: String) -> Unit> = mutableListOf(
        { tag, message -> android.util.Log.i(tag, message) }
    )

    fun e(tag: String, message: String, cause: Throwable? = null) {
        if (level < Level.Error) return
        errorHandlers.forEach { it(tag, message, cause) }
    }

    fun w(tag: String, message: String) {
        if (level < Level.Warn) return
        warningHandlers.forEach { it(tag, message) }
    }

    // Info a message. Using lambda avoids computing the message if log level is not enabled. Useful for messages that are costly to produce
    fun i(tag: String, messageHandler: (() -> String)) {
        if (level < Level.Info) return
        infoHandlers.forEach { it(tag, messageHandler()) }
    }

    fun i(tag: String, message: String) {
        if (level < Level.Info) return
        infoHandlers.forEach { it(tag, message) }
    }

    // Debug a message. Using lambda avoids computing the message if log level is not enabled. Useful for messages that are costly to produce
    fun d(tag: String, messageHandler: (() -> String)) {
        if (level < Level.Debug) return
        android.util.Log.d(tag, messageHandler())
    }

    fun d(tag: String, message: String) {
        if (level < Level.Debug) return
        android.util.Log.d(tag, message)
    }
}
